### PR TITLE
Add optimizer 

### PR DESCRIPTION
**Pull request by yixinglu**
depends on #240 

This PR has not consider cost calculation of each plan node which has equal default cost 0.0. And I will continue to implement more better strategy to select best plan. 

By given a `PushFilterDownGetNbrsRule` example, others could add more rules like that and test the rule by comparing the optimized plan with expected result in `test_optimizer.py`.

This's one tiny step for optimization of Graph Query engine. More works are to be continued.

### CODE REVIEW

File: src/optimizer/Optimizer.cpp [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+Status Optimizer::doExploration() {
+    // TODO(yee): Apply all rules recursively, not only once round
+    for (auto rule : ruleSet_->rules()) {
+        if (!rootGroup_->isExplored(rule)) {
+            NG_RETURN_IF_ERROR(rootGroup_->explore(rule));
+        }
+    }
+    return Status::OK();
+}
+
```
Comment (bright-starry-sky) [inline src/optimizer/Optimizer.cpp:54]: How to set the default rule when `enable_optimizer=false`?

Comment (yixinglu) [reply src/optimizer/Optimizer.cpp:54]: I'll control rule set selection in query instance. pls wait a minute.

Comment (yixinglu) [reply src/optimizer/Optimizer.cpp]: Now you can add opt rule to `RuleSet::defaultRules()` like [PushFilterDownGetNbrsRule](https://github.com/vesoft-inc/nebula-graph/pull/247/files#diff-1628cc036fa9a70c831e51ce2cb0884fR32)

File: src/optimizer/rule/PushFilterDownGetNbrsRule.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+namespace opt {
+
+class PushFilterDownGetNbrsRule final : public OptRule {
+public:
+    static std::unique_ptr<OptRule> kInstance;
+
+    bool match(const OptGroupExpr *groupExpr) const override;
+    Status transform(graph::QueryContext *qctx,
+                     const OptGroupExpr *groupExpr,
+                     TransformResult *result) const override;
```
Comment (bright-starry-sky) [inline src/optimizer/rule/PushFilterDownGetNbrsRule.h:28]: A question, Way `const` of method transform? Doing so will not allow set private variables in class xxxxRule. WDYT?

**Review by dutor — APPROVED** [review]
LGTM.

**Comment by bright-starry-sky** [comment]
Please resolve the conflict. Thanks.

File: src/optimizer/rule/PushFilterDownGetNbrsRule.h (hunk 2 above)
Comment (yixinglu) [reply src/optimizer/rule/PushFilterDownGetNbrsRule.h:28]: Each rule will be used by many queries, and we register each rule in the singleton way into rule set. so change state of each rule will affect optimization of other queries. we should keep each rule stateless.

**Comment by yixinglu** [comment]
@bright-starry-sky @dutor resolved conflicts.

**Review by bright-starry-sky — APPROVED** [review]
LGTM.

**Review by dutor — APPROVED** [review]
LGTM.

**Pull request merged**